crypto/bn/bn_mpi.c

```diff
@@ -45,7 +45,7 @@ BIGNUM *BN_mpi2bn(const unsigned char *d, int n, BIGNUM *ain)
     int neg = 0;
     BIGNUM *a = NULL;
 
-    if (n < 4) {
+    if (n < 4 || d[0] & 0x80) {
         BNerr(BN_F_BN_MPI2BN, BN_R_INVALID_LENGTH);
         return NULL;
     }
```

Comment:
Oouch, the line below is
```C
len = ((long)d[0] << 24) | ((long)d[1] << 16) | ((int)d[2] << 8) | (int)
    d[3];
```
This was apparently written in the days when int was a 16 bit short value.
So if you like please rewrite the whole stuff, in a PR of its own.

Reply:
The line below is technically correct albeit a little bit ugly.